Implemented STU3

jarimayenburg opened the pull request:
Great library! We needed STU3 support so I used your generator to generate the STU3 model and added an STU3 client with the appropriate tests. All tests should be passing, let me know what you think.

FlixCoder commented:
Thanks, I will need some time to look up the STU3 spec and such and check everything, as it is just my free time :D

jarimayenburg commented:
Take all the time you need, we will be using our fork until this is merged anyway

FlixCoder reviewed (commented):
Alright, found time quite early actually, so was way faster.

I still need to check a few things, but it already looks quite good! I mean much of it is just a copy (still need to find a way to get rid of the code duplication, any nice ideas?).

Only missing checks:
- Completeness of test examples
- Correctness of JSON schema
- Generated code matches the current generate crate
- xtask test actually runs through

So everything should be fine ^^

A few remarks I have:

FlixCoder commented:
Huh, apparently you aren't using the most up to date definitions of STU3. Do you mind if I switch to the newest downloads version? Do I have access to push here? :D I would of course only add a commit so your contribution would remain the same.

There is also (also done on my local branch):
- Client needs a method `stu3` to switch the client to STU3 mode.
- OCD kicking in with STU3 not being in front of R4B in some cases :P 
- VSCode settings need to enable STU3 feature to show

jarimayenburg commented on crates/fhir-sdk/src/client/mod.rs:
```diff
@@ -36,13 +38,19 @@ pub struct FhirR4B;
 /// Default client version.
 type DefaultVersion = FhirR5;
-#[cfg(all(not(feature = "r5"), feature = "r4b"))]
+#[cfg(all(not(feature = "r5"), not(feature = "stu3"), feature = "r4b"))]
```
Sounds good!

jarimayenburg commented on crates/generate/src/generate/gen_types.rs:
```diff
@@ -261,7 +261,11 @@ fn generate_choice_field(
-	let variants = field.types.iter().map(|ty| {
+	let mut types = field.types.clone();
+
+	types.dedup();
```
You are right, that would be much better. Let me fix that.

jarimayenburg replied:
I moved it now. The reason that this change is needed is that in STU3, `ElementDefinition.type.targetProfile` has cardinality `0..1` in STU3, while in R4 and up `ElementDefinition.type.targetProfile` has cardinality `0..*`. So in STU3, an element that can be of type "Reference" with multiple allowed target profiles is represented in `ElementDefinition.type` as multiple entries with the same `code` "Reference". Without this deduplication, the code generator would generate enums with multiple variants called "Reference".

FlixCoder replied:
Ah, ok, thanks

FlixCoder commented on crates/generate/Cargo.toml:
```diff
@@ -6,7 +6,7 @@ version = "0.1.0"
 Inflector = "0.11.4"
 anyhow = "1.0.70"
-fhir-model = {version = "0.6.0", default-features = false, features = ["r4b", "r5"]}
+fhir-model = {path = "../fhir-model", default-features = false, features = ["stu3", "r4b", "r5"]}
```
No lets keep it, I will do it once I released a new version with stu3

jarimayenburg commented:
> Huh, apparently you aren't using the most up to date definitions of STU3. Do you mind if I switch to the newest downloads version? Do I have access to push here? :D I would of course only add a commit so your contribution would remain the same.

The version of STU3 I used was 3.0.2, released on Feb 21, 2017. That should be the latest one right? Like you, I had to edit a bunch of things in the spec JSON files to get it working. Some things I changed:
1. Added `experimental`, `version` and `fhirVersion` fields to StructureDefinitions and CodeSystems wherever they were missing.
2. Added `code` fields to ElementDefinitions of primitive datatypes that only had a `_code` extension.

> * Client needs a method `stu3` to switch the client to STU3 mode.
Let me add that!

> * OCD kicking in with STU3 not being in front of R4B in some cases :P
Haha I had the same OCD kicking in. Problem is that `rustfmt` orders modules alphabetically, so `stu3` will land after `r4b` and `r5`. Feel free to order it however you'd like when it's merged though.

> * VSCode settings need to enable STU3 feature to show
Whoops, I don't use VSCode so I missed that. Added it now.

FlixCoder commented:
> he version of STU3 I used was 3.0.2, released on Feb 21, 2017. That should be the latest one right?

Hmm the `version.info` file says it is 3.0.1 :D And the newest version is a build of `3.0.2` from 2019 apparently.

> Haha I had the same OCD kicking in. Problem is that rustfmt orders modules alphabetically, so stu3 will land after r4b and r5. Feel free to order it however you'd like when it's merged though.

No I am using rustfmt too ^^ So I will need to live with it :D


Alright, I think that's it, I need to take a final look and then I will merge it, thank you very much!

jarimayenburg commented:
> > he version of STU3 I used was 3.0.2, released on Feb 21, 2017. That should be the latest one right?
> 
> Hmm the `version.info` file says it is 3.0.1 :D And the newest version is a build of `3.0.2` from 2019 apparently.

On the version history page (https://hl7.org/fhir/history.html), in that big table, I think the correct latest version is called "3.0.0" and released on Mar 21, 2017. While at the top of the page they refer to Release 3 STU as 3.0.2, released on Feb 21, 2017. Not sure how a newer version 3.0.2 has an earlier release date than the older version 3.0.0, but here we are.

Anyway, I used the link to 3.0.2 in the "major milestones" to get to the final STU3 spec and got the specification from the download page of that version: https://hl7.org/fhir/stu3/downloads.html. I can't imagine that that is an older version.

FlixCoder commented:
Alright, then it should be the newest :D Weird. No, totally fine then, thanks!
